Add release notes for 0.257 (#16399)
prestodb/presto: presto-docs/src/main/sphinx/release/release-0.257.rst

=============
Release 0.257
=============

**Details**
===========

General Changes
_______________

> * Fix query exceeding local memory limit when unspilling data due to incorrect memory size calculation.

Reviewer: oh yeah that's better

> * Fix query exceeding local memory limit when unspilling data due to incorrect memory size calculation.

Author: done

> * Fix deadlock for queries with join and limit when spilling is enabled.

Author: done

> * Fix expressions with multiple ``IS NULL`` on the same variable.

Author: Updated to what Mayank suggested for now.

> * Fix query failures due to expressions with multiple ``IS_NULL`` on the same variable.

Reviewer 2: The way it's written, it's as if `IS_NULL` is a function.  Let's remove the underscore and make it lower case to make it clearer. 

```suggestion
* Fix query failures due to expressions with multiple ``is null`` checks on the same variable.
```

Author: done

* Improve memory usage of queries spilling in the join operator.

> * Improve plan expressions to be closer to valid SQL.

Reviewer 3: `plan expressions` makes me think that this is some kind of an expression. So lets clarify that we are talking about expressions printed in plans.

```
Improve expressions printed in query plans to be closer to valid SQL.
```

Author: done

* Add support for instance count in function :func:`array_position`.

> * Add coercion support when correlated symbols are in non-direct column reference type.

Reviewer: This is more confusing to me, and also in the end this PR added support for all kinds of expressions, not just casts.  Maybe say

`Add support for correlated subqueries with complex expressions in the correlation`

Author: done

> * Add support for the ``OFFSET`` clause in SQL query expressions. This feature can be enabled by setting the session property ``offset_clause_enabled`` or configuration property ``offset-clause-enabled`` to true.

Reviewer 3: ```
``true``
```

Author: done

=============
Release 0.257
=============

**Details**
===========

General Changes
_______________

> * Fix scheduling issue in join spill to improve memory used by unspilled data.

Reviewer: Improve memory usage of queries spilling in the join operator.

Author: done

> * Fix spilled query exceeding local memory limit due to incorrect memory size calculation when spilled data is read by SingleStreamSpiller.

Reviewer: Fix query exceeding local memory limit when unspilling data due to incorrect memory size calculation.

Author: done

> * Fix deadlock problem that results into join + limit getting stuck when spilling is enabled.

Reviewer: Fix deadlock for queries with join and limit when spilling is enabled.

Author: done

> * Improve RowExpressionFormatter output to be closer to valid SQL. Constant expressions are now output as ``TYPE 'value'`` (e.g. ``BIGINT'2'``) using TypeConstructor (with single quotes appearing in the string escaped with ``''``), with the exception of ``VARBINARY`` constants, which use the binary literal syntax (e.g. ``X'abc123'``). ``LIKE`` expressions are now output as valid SQL.

Reviewer 2: We should not refer to class names here and try to frame it in terms of the functionality being change.  Perhaps this would work:

```suggestion
* Improve plan expressions to be closer to valid SQL. 
```

Author: done

> * Add support for instance count in function :func:`array_position`.

Reviewer 2: Link to documentation?

Author: The generated html will link to the documentation. See 
![image](https://user-images.githubusercontent.com/236188/124998602-8ff94800-e001-11eb-8e76-8b8a93050a09.png)
Hope that answers your question.

Reviewer 3: `instance count` is not clear to me. Maybe

```
Add support to find the n-th instance in :func:`array_position`.
```

Author: done

> * Add support for expressions with multiple ``IS NULL`` on the same variable.

Reviewer 2: ```suggestion
* Fix expressions with multiple ``IS NULL`` on the same variable.
```

Author: done

> * Add support for more kinds of correlated subqueries.

Reviewer 2: Should we enumerate the types of subqueries?

Author: Updated to `* Add coercion support when correlated symbols are in non-direct column reference type.`

> * Add support for the result offset clause in SQL query expressions. This feature can be enabled by setting the session property ``offset_clause_enabled`` or configuration property ``offset-clause-enabled`` to true.

Reviewer: what is the result offset clause?

Author: I think it is just the `OFFSET` clause. See https://github.com/prestodb/presto/pull/16035
I dropped result to make it more clear.

* Add support for multi coordinator in ``v1/resourceGroupState/*`` endpoint.

> * Add coordinator host name in web ui to support multi coordinator setup.

Reviewer 2: I would remove these lines since they're still beta.  We can add a line for the broader feature once it's GA.
